src/checkpointd/runtime.rs

```diff
@@ -0,0 +1,336 @@
+}
+
+pub struct Runtime {
+    identity: ServiceId,
+    checkpoints: CheckpointGetSet,
+    pending_checkpoint_chunks: HashMap<[u8; 20], HashSet<CheckpointChunk>>,
+}
+
+impl Runtime {}
+
+impl CtlServer for Runtime {}
+
+impl esb::Handler<ServiceBus> for Runtime {
+    type Request = Request;
+    type Error = Error;
+
+    fn identity(&self) -> ServiceId {
+        self.identity.clone()
+    }
+
+    fn handle(
+        &mut self,
+        endpoints: &mut Endpoints,
+        bus: ServiceBus,
+        source: ServiceId,
+        request: Request,
+    ) -> Result<(), Self::Error> {
+        match bus {
+            ServiceBus::Msg => self.handle_rpc_msg(endpoints, source, request),
+            ServiceBus::Ctl => self.handle_rpc_ctl(endpoints, source, request),
+            _ => Err(Error::NotSupported(ServiceBus::Bridge, request.get_type())),
+        }
+    }
+
+    fn handle_err(&mut self, _: &mut Endpoints, _: esb::Error<ServiceId>) -> Result<(), Error> {
+        // We do nothing and do not propagate error; it's already being reported
+        // with `error!` macro by the controller. If we propagate error here
+        // this will make whole daemon panic
+        Ok(())
+    }
+}
+
+impl Runtime {
+    fn handle_rpc_msg(
+        &mut self,
+        _endpoints: &mut Endpoints,
+        _source: ServiceId,
+        request: Request,
+    ) -> Result<(), Error> {
+        match request {
+            Request::Hello => {
+                // Ignoring; this is used to set remote identity at ZMQ level
+            }
+
+            req => {
+                error!(
+                    "MSG RPC can only be used for forwarding farcaster protocol messages, found {:?}, {:#?}",
+                    req.get_type(), req
+                )
+            }
+        }
+        Ok(())
+    }
+
+    fn handle_rpc_ctl(
+        &mut self,
+        endpoints: &mut Endpoints,
+        source: ServiceId,
+        request: Request,
+    ) -> Result<(), Error> {
+        match request {
+            Request::Hello => match &source {
+                source => {
+                    debug!("Received Hello from {}", source);
+                }
+            },
+
+            Request::CheckpointMultipartChunk(request::CheckpointMultipartChunk {
+                checksum,
+                msg_index,
+                msgs_total,
+                serialized_state_chunk,
+                swap_id,
+            }) => {
+                debug!("received checkpoint multipart message");
+                if self.pending_checkpoint_chunks.contains_key(&checksum) {
+                    let chunks = self
+                        .pending_checkpoint_chunks
+                        .get_mut(&checksum)
+                        .expect("checked with contains_key");
+                    chunks.insert(CheckpointChunk {
+                        msg_index,
+                        serialized_state_chunk,
+                    });
+                } else {
```

Comment:
I don't understand what you are asking here. Either it adds to the chunks if a chunk with the key already exists, or it creates a new entry for that chunk if a key does not exist yet.
I also don't understand what you mean with "zmq does it simpler". Do you mean I should use the msg length instead of its index? What if two messages are the same size?

Reply:
you'd have to add extra headers to the message, i'd guess:

- length
- checksum
- serviceid? (depending on the socket type)

Reply:
But then send it as a raw zmq message to the databased runtime, that knows how to decode those messages correctly

Reply:
That's exactly what is in the message right now. Length is provided by strict encoding, and serviceid and checksum are part of the message struct.

Reply:
but that you're doing over the service bus, what im saying is do it over zmq directly